pyfaidx/__init__.py

```diff
@@ -359,31 +383,24 @@ def __init__(self,
                 raise ImportError(
                     "BioPython >= 1.73 must be installed to read block gzip files.")
             else:
-                self._fasta_opener = bgzf.open
                 self._bgzf = True
-        elif filename.lower().endswith('.bz2') or filename.lower().endswith(
-                '.zip'):
+                try:
+                    # mutable mode is not supported for bzgf anyways
+                    self.file = bgzf.BgzfReader(fileobj=self.file, mode="b")
+                except (ValueError, IOError):
+                    raise UnsupportedCompressionFormat(
+                        "Compressed FASTA is only supported in BGZF format. Use "
+                        "the samtools bgzip utility (instead of gzip) to "
+                        "compress your FASTA."
+                    )
+        elif self.filename.lower().endswith(('.bz2', '.zip')):
             raise UnsupportedCompressionFormat(
                 "Compressed FASTA is only supported in BGZF format. Use "
                 "bgzip to compresss your FASTA.")
         else:
-            self._fasta_opener = open
             self._bgzf = False
 
-        try:
-            self.file = self._fasta_opener(filename, 'r+b'
-                                           if mutable else 'rb')
-        except (ValueError, IOError) as e:
-            if str(e).find('BGZF') > -1:
-                raise UnsupportedCompressionFormat(
-                    "Compressed FASTA is only supported in BGZF format. Use "
-                    "the samtools bgzip utility (instead of gzip) to "
-                    "compress your FASTA.")
-            else:
-                raise FastaNotFoundError(
-                    "Cannot read FASTA file %s" % filename)
-
-        self.indexname = filename + '.fai'
+        self.indexname = self.filename + '.fai'
```

Comment:
Good point.
Currently this stores the index file on the remote.
So: providing `fsspec.open("s3://mybucket/somefile.fasta")` will make `self.filename = "mybucket/somefile.fasta"` and `self._fs` will be an s3fs filesystem instance. `self.indexname = mybucket/somefile.fasta.fai`.

But this requires the remote to be writable. (I basically stuck to the current behavior that stores the index next to the file.) In case the remote is readonly, it would be nice if the user could specify a local path for the index file.

I could make this configurable through a kwarg, with the default being current behavior.

Reply:
I think the current behavior is fine, and as I haven't used fsspec I did not understand that the remote file-like object would be created. As long as there is as reasonable exception generated when the remote is read-only.

Reply:
I'll add a test for this on read-only filesystem with a fasta file with no index file.

Reply:
> In case the remote is readonly, it would be nice if the user could specify a local path for the index file.

I ran into this when experimenting with an implementation a little while ago. It's hard to know what the "correct" behavior should be. I guess one option would be to have an `index_filename` kwarg in the constructor, which could similarly be a `fsspec.OpenFile` or `str` and would specify explicitly where to look for (or create) an index.

So the default would be to read and write index adjacent to `filename` in whatever filesystem is provided. If `filename` is readonly, then we could raise an exception, and the user could fix either by 1.) configuring the filesystem with write permissions _or_ 2.) specifying a local path (e.g., `./local.fasta.fai`). My only concern is that it could get challenging to handle multiple file objects.